### PR TITLE
Improve Test_DataVector

### PR DESCRIPTION
**Pull request by fmahebert**
## Proposed changes

This PR aims to organize the tests cases within Test_DataVector, which were a little scattered and sometimes redundant. In particular, the changes done are,

1. Split the long "Math" test to pull out the testing of reference DataVectors and `array<DataVector>`.
1. Group the checks in the "Math" test according to the type of expression being tested.
1. Add a few more checks of math with reference DataVectors, with the intention of closing #83. Perhaps more checks are still needed, so let me know if you can think of any I should add.
1. Add some scoping and comments throughout the file to improve legibility.
1. Remove some redundant checks within/between test cases.
1. Reorder some tests (purely to satisfy my sense of aesthetics).

Still to do:

1. Squash these commits.
1. Understand why clang-format chokes on the 2nd test case ("DataVector_Ref") and produces garbage formatting.

### Types of changes:

- [x] Cleanup
- [ ] Bugfix
- [x] New feature

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] Follows [code review guidelines](https://sxs-collaboration.github.io/spectre/code_review_guide.html)
- [ ] Code has documentation and unit tests
- [ ] Private member variables have a trailing underscore
- [ ] Do not use [Hungarian notation](https://en.wikipedia.org/wiki/Hungarian_notation), e.g. `double* pd_blah` is bad
- [ ] Header order:
  1. hpp corresponding to cpp (only in cpp files)
  2. Blank line (only in cpp files)
  3. STL and externals (in alphabetical order)
  4. Blank line
  5. SpECTRE includes (in alphabetical order)
- [ ] File lists in CMake are alphabetical
- [ ] Correct `noexcept` specification for functions (if unsure, mark `noexcept`)
- [ ] Mark objects `const` whenever possible
- [ ] Almost always `auto`, except with expression templates, i.e. `DataVector`
- [ ] All commits for performance changes provide quantitative evidence and the tests used to obtain said evidence.
- [ ] Make sure error messages are helpful, e.g. "The number of grid points in the matrix 'F' is not the same as the number of grid points in the determinant."

### CODE REVIEW

**Review by nilsdeppe — CHANGES_REQUESTED** [review]
Declined, doesn't add any benefit to the code.

File: tests/Unit/DataStructures/Test_DataVector.cpp [hunk 1]
```diff
@@ -173,53 +173,45 @@ void check_vectors(const T1& t1, const T2& t2) {
...
+    b = m0 + m1;
+    CHECK(a.size() == 0);  // NOLINT
+    check_vectors(b, DataVector(10, 12.0));
     a = m0 * m1;
     check_vectors(a, DataVector(10, 27.0));
     check_vectors(b, DataVector(10, 12.0));
   }
 
-  {
-    DataVector a(10, 2.0);
```
Comment (wthrowe) [inline tests/Unit/DataStructures/Test_DataVector.cpp]: Please revert these changes.  These tests were not redundant, but were testing for a regression of a bug involving moving erroneously causing a partial aliasing of DataVectors.  In particular, the move from `a` to `b` followed by an expression template assignment to `a` with no intervening modification of `b` is important to the test.

File: tests/Unit/DataStructures/Test_DataVector.cpp [hunk 2]
```diff
@@ -171,6 +171,223 @@ void check_vectors(const T1& t1, const T2& t2) {
...
+
+    check_vectors(step_function(DataVector{-12.3, 2.0, -4.0, 0.0, 7.0, -8.0}),
+                  DataVector{0.0, 1.0, 0.0, 1.0, 1.0, 0.0});
+  }
+
+  // Test composition of constant expressions with DataVector math member
+  // functions
+  {
+    DataVector x(num_pts, 2.);
+    check_vectors(DataVector(num_pts, 0.82682181043180603), square(sin(x)));
```
Comment (wthrowe) [inline tests/Unit/DataStructures/Test_DataVector.cpp:278]: While you're cleaning this up, these tests could calculate the correct values instead of using hardcoded answers.

**Review by wthrowe — CHANGES_REQUESTED** [review]

File: tests/Unit/DataStructures/Test_DataVector.cpp [hunk 3]
```diff
@@ -392,8 +393,10 @@ SPECTRE_TEST_CASE("Unit.DataStructures.DataVector.Math",
   x = DataVector(num_pts, 2.);
   check_vectors(DataVector(num_pts, 0.82682181043180603), square(sin(x)));
   check_vectors(DataVector(num_pts, -0.072067555747765299), cube(cos(x)));
+}
 
-  // Test addition of arrays of DataVectors to arrays of doubles.
+SPECTRE_TEST_CASE("Unit.DataStructures.DataVector.Math_array<DataVector>",
```
Comment (nilsdeppe) [inline tests/Unit/DataStructures/Test_DataVector.cpp]: I like splitting of the test, I think this is a good thing to keep

File: tests/Unit/DataStructures/Test_DataVector.cpp [hunk 4]
```diff
@@ -410,40 +413,46 @@ SPECTRE_TEST_CASE("Unit.DataStructures.DataVector.Math",
   const DataVector e11{-20.0, -20.0, -20.0, -20.0};
   const DataVector e12{-30.0, -30.0, -30.0, -30.0};
 
-  const std::array<double, 3> point{{10.0, 20.0, 30.0}};
```
Comment (nilsdeppe) [inline tests/Unit/DataStructures/Test_DataVector.cpp:413]: I don't like scoping (personally, but that's not all too important). The difficulty is that we have to check every single line very carefully, ideally with a script to make sure they are the same. This puts a lot of burden on the reviewers. If there was something missing, it would be useful to add.

File: tests/Unit/DataStructures/Test_DataVector.cpp [hunk 5]
```diff
@@ -410,40 +413,46 @@ SPECTRE_TEST_CASE("Unit.DataStructures.DataVector.Math",
...
+    }
+    array0 -= array1;
+    array0 -= array1;
+    for (size_t i = 0; i < 3; i++) {
+      check_vectors(gsl::at(array0, i), gsl::at(expected4, i));
+    }
   }
 
-  // Test calculation of magnitude of DataVector
+  // Test calculation of magnitude for array<DataVector>
```
Comment (nilsdeppe) [inline tests/Unit/DataStructures/Test_DataVector.cpp:490]: This clarification is fine

File: tests/Unit/DataStructures/Test_DataVector.cpp [hunk 6]
```diff
@@ -48,25 +48,6 @@ SPECTRE_TEST_CASE("Unit.DataStructures.DataVector", "[DataStructures][Unit]") {
   CHECK(t_move_constructor.is_owning());
 }
 
-SPECTRE_TEST_CASE("Unit.Serialization.DataVector",
```
Comment (nilsdeppe) [inline tests/Unit/DataStructures/Test_DataVector.cpp]: Why move this at all? I usually just use search to find the test I need to edit.

File: tests/Unit/DataStructures/Test_DataVector.cpp [hunk 7]
```diff
@@ -10,42 +10,53 @@
 #include "tests/Unit/TestHelpers.hpp"
 
 SPECTRE_TEST_CASE("Unit.DataStructures.DataVector", "[DataStructures][Unit]") {
```
Comment (nilsdeppe) [inline tests/Unit/DataStructures/Test_DataVector.cpp:12]: Please don't just add scopes. It leads to horrible diffs and requires the reviewers to very carefully review every single line, and think about the change as a whole to make sure the efficacy of the test is unaltered.

File: tests/Unit/DataStructures/Test_DataVector.cpp [hunk 8]
```diff
@@ -61,57 +61,74 @@ SPECTRE_TEST_CASE("Unit.DataStructures.DataVector", "[DataStructures][Unit]") {
 
 SPECTRE_TEST_CASE("Unit.DataStructures.DataVector_Ref",
                   "[DataStructures][Unit]") {
-  DataVector data{1.43, 2.83, 3.94, 7.85};
```
Comment (nilsdeppe) [inline tests/Unit/DataStructures/Test_DataVector.cpp:71]: Similarly to before, don't just add scopes everywhere, I have no idea if the changes you made do indeed leave the test unaltered.

File: tests/Unit/DataStructures/Test_DataVector.cpp [hunk 9]
```diff
@@ -173,53 +173,45 @@ void check_vectors(const T1& t1, const T2& t2) {
 
 SPECTRE_TEST_CASE("Unit.DataStructures.DataVector.MathAfterMove",
                   "[Unit][DataStructures]") {
```
Comment (nilsdeppe) [inline tests/Unit/DataStructures/Test_DataVector.cpp]: As Will pointed out, things that are seemingly redundant may not be with DataVector. The internals are keeping track of owning-ness in non-trivial ways and so altering the tests is something I frown upon. A fair number of the tests were added as a result of bugs and so altering them could allow re-introduction of said bug.

File: tests/Unit/DataStructures/Test_DataVector.cpp [hunk 10]
```diff
@@ -171,6 +171,223 @@ void check_vectors(const T1& t1, const T2& t2) {
 }
 }  // namespace
 
+SPECTRE_TEST_CASE("Unit.DataStructures.DataVector.Math",
```
Comment (nilsdeppe) [inline tests/Unit/DataStructures/Test_DataVector.cpp:174]: I simply cannot review this. We need to formally verify that every change is a no-op even with regard to the potential non-owningness of DataVector.

File: tests/Unit/DataStructures/Test_DataVector.cpp [hunk 11]
```diff
@@ -171,6 +171,223 @@ void check_vectors(const T1& t1, const T2& t2) {
...
+                  "[Unit][DataStructures]") {
+  // Test min/max/abs
```
Comment (nilsdeppe) [inline tests/Unit/DataStructures/Test_DataVector.cpp:176]: I disagree that this is cleaning anything up. The code was already sectioned by functions it was testing, all this does it add a lot of work for reviewers with zero benefit. If you're searching for the portion that tests `exp` you're just going to search for `exp`, you don't need additional scope and a comment to tell you where to look.

**Review by nilsdeppe — CHANGES_REQUESTED** [review]
I've added specific comments addressing what I don't like. Hopefully that'll help.

**Comment by fmahebert** [comment]
Fair enough — I wasn't making these changes with review-ability in mind. I will tone it down and resubmit.
